File: Heap/min_heap.cpp
// Min Heap implementation using arrays
#include <iostream>
#include <vector>

using namespace std;

struct MinHeap {
    
    int heapSize = 0;
    int noOfElements = 0;
    vector<int> heap;
    
    MinHeap(int size) {    
        heapSize = size;
        heap.resize(heapSize+1);
        heap.shrink_to_fit();
        heap[0] = 0;
    }

    bool add(int element) {
        
        if(noOfElements + 1 > heapSize) {
            return false;
        }
        noOfElements++;

        heap[noOfElements] = element;

        int currIndex = noOfElements;
        int parentIndex = currIndex / 2;

        while(heap[currIndex] < heap[parentIndex] && currIndex > 1) {
            swap(heap[currIndex], heap[parentIndex]);
            currIndex = parentIndex;
            parentIndex = currIndex / 2;
        }

        return true;
    }

    bool pop() {

        if(noOfElements < 1) {
            return false;
        }

        heap[1] = heap[noOfElements];

        noOfElements--;

        int currIndex = 1;

        while(currIndex <= noOfElements/2) {
            int left = currIndex * 2;
            int right = (currIndex * 2) + 1;
            
            if(heap[currIndex] > heap[left] || heap[currIndex] > heap[right]) {
                if(heap[left] < heap[right]) {
                    int temp = heap[left];
                    heap[left] = heap[currIndex];
                    heap[currIndex] = temp;
                    currIndex = left;
                } else {
                    int temp = heap[right];
                    heap[right] = heap[currIndex];
                    heap[currIndex] = temp;
                    currIndex = right;
                }
            } else {
                break;
            }
        }
        return true;
    }

    int peek() {
        return heap[1];
    }

    int size() {
        return noOfElements;
    }
};

int main() {
    
    MinHeap *minHeap = new MinHeap(3);

    cout<<"[add]: "<<minHeap->add(3)<<endl;
    cout<<"[add]: "<<minHeap->add(1)<<endl;
    cout<<"[add]: "<<minHeap->add(2)<<endl;

    cout<<"[top]: "<<minHeap->peek()<<endl;

    cout<<"[pop]: "<<minHeap->pop()<<endl;

    cout<<"[top]: "<<minHeap->peek()<<endl;

    cout<<"[size]: "<<minHeap->size()<<endl;

    return 0;
}

File: Heap/min_heap.rb
class MinHeap
  def initialize
    @heap = []
  end

  def insert(val)
    @heap << val
    heapify_up(@heap.size-1)
  end

  def extract
    min_val = @heap.first
    @heap[0] = @heap.pop
    heapify_down(0)
    return min_val
  end

  def peek
    @heap.first
  end

  def empty?
    @heap.empty?
  end

  private

  def heapify_up(index)
    parent_index = (index-1)/2
    if index > 0 && @heap[index] < @heap[parent_index]
      @heap[index], @heap[parent_index] = @heap[parent_index], @heap[index]
      heapify_up(parent_index)
    end
  end

  def heapify_down(index)
    child_index = 2 * index + 1
    if child_index < @heap.size
      right_child_index = child_index+1
      if right_child_index < @heap.size && @heap[right_child_index] < @heap[child_index]
        child_index = right_child_index
      end
      if @heap[child_index] < @heap[index]
        @heap[index], @heap[child_index] = @heap[child_index], @heap[index]
        heapify_down(child_index)
      end
    end
  end
end

min_heap = MinHeap.new
min_heap.insert(10)
min_heap.insert(15)
min_heap.insert(20)
min_heap.insert(17)
min_heap.insert(8)

puts "Peek min: #{min_heap.extract}"
puts "Peek min: #{min_heap.extract}"
puts "Peek min: #{min_heap.extract}"
puts "Peek min: #{min_heap.extract}"
puts "Peek min: #{min_heap.extract}"
